### ISSUE TITLE
`show_default`, but without type parameters

### ISSUE THREAD

**Issue by goretkin** [JuliaLang/julia#36263]
Definitions like https://github.com/JuliaLang/julia/blob/1e761112038518bafe6ce37b646ae38f38d28253/base/range.jl#L742 could be no more than something like

```julia
show(io::IO, r::OneTo) = show_default(io, r; type_parameters=false)
```

There are many places where that's a default behavior that is wanted. For example, when all the type parameters correspond to fields that when `show`'d reveal their type. For example

```julia
struct Foo{T}
  x::T
end

show(Foo(Base.OneTo(3)))
```

produces

`Foo{Base.OneTo{Int64}}(Base.OneTo(3))`,
which is unnecessary.

Also, `show_default` has great logic to use `IOContext` to see which names need to be qualified or not:

```julia
julia> Base.show_default(stdout, Base.OneTo(3))
Base.OneTo{Int64}(3)
julia> Base.show_default(stdout, Foo(Base.OneTo(3)))
Foo{Base.OneTo{Int64}}(Base.OneTo(3))
julia> using Base: OneTo

julia> Base.show_default(stdout, Base.OneTo(3))
OneTo{Int64}(3)
julia> Base.show_default(stdout, Foo(Base.OneTo(3)))
Foo{OneTo{Int64}}(Base.OneTo(3))
```

Notice the inconsistency in `Foo{OneTo{Int64}}(Base.OneTo(3))`. 

x-ref: https://github.com/JuliaLang/julia/issues/33260

**Comment by goretkin** [comment]
Seems suspiciously relevant
https://github.com/JuliaLang/julia/blob/3222aba1d32ab9af41c82a1dd0bce0a7f72d8cd5/base/show.jl#L477-L483 but doesn't do the job fully.

One idea is a trait-based API, where a type could register "when you `show` me, my type is obvious", but that would require more careful design since it would likely be public.

**Comment by goretkin** [comment]
Related: https://github.com/MrVPlusOne/SimpleTypePrint

**Comment by timholy** [comment]
If a change like this is made, it shouldn't be done without an awareness that at least until recently `show_default` was a major invalidation risk. One would want to run the tests in #37193, which had to be reverted (see #38461), to make sure this kind of change doesn't cause unintended problems.

**Comment by timholy** [comment]
Also,
```julia
julia> r = Base.OneTo(3)
Base.OneTo(3)

julia> io = IOBuffer();

julia> @btime show($io, $r)
  135.656 ns (2 allocations: 96 bytes)

julia> @btime Base.show_default($io, $r)
  345.577 μs (44 allocations: 21.73 KiB)
```
A 2000x performance regression is not something to embrace lightly.

**Comment by bramtayl** [comment]
Is there a way to make `show_default` more performant?

**Comment by goretkin** [comment]
> A 2000x performance regression is not something to embrace lightly.

I agree. I suspect this is due to "show_default has great logic to use `IOContext` to see which names need to be qualified or not" and it also has logic to print circular references.

I notice that the definitions prevent specializing compiled code to the type of `x`: https://github.com/JuliaLang/julia/blob/814945c34f125e12f66c37605b67eb348d4b9cd2/base/show.jl#L390-L397

It seems reasonable that specializing on the type would abate that performance penalty, at the cost of introducing a compilation penalty.

**Comment by timholy** [comment]
> It seems reasonable that specializing on the type would abate that performance penalty

When the type is known to inference, yes. When the type is not known to inference (e.g., printing elements of a `Vector{Any}`) it would have precisely the opposite effect. Currently

```julia
julia> using MethodAnalysis

julia> methodinstances(Base._show_default)
3-element Vector{Core.MethodInstance}:
 MethodInstance for _show_default(::IOContext{IOBuffer}, ::Any)
 MethodInstance for _show_default(::IOBuffer, ::Any)
 MethodInstance for _show_default(::IO, ::Any)
```
so you only have to dispatch on the `io` argument. If you allow specialization on the type, now you'll go from 3 MethodInstances to hundreds or thousands, and now you have to runtime dispatch on the type argument, and that will be bad for runtime performance. Given that it would also murder latency (it's an expensive method to compile because it's complicated), I think it's very unlikely that this would be a win overall. This is a textbook case for `@nospecialize`.

**Comment by timholy** [comment]
> I suspect this is due to "show_default has great logic to use IOContext to see which names need to be qualified or not" and it also has logic to print circular references.

Don't guess, profile. 99% of the time is spent on this line: https://github.com/JuliaLang/julia/blob/a4cd68cceb78636fd9968e8464a7e12cd333bec4/base/show.jl#L399

It ends up spending all of its time in calling `show_typealias`. Most of its time is spent looking up names in modules and deciding whether there's a corresponding `isdefined` object.
